Fix several bugs causing the page to reflow on every mouse move event

pcwalton opened the pull request:
After all these changes are applied, Hacker News and GitHub only repaint and reflow nodes that actually have hover styles applied when the mouse moves over them.

r? @mbrubeck 

cc @bholley

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/8299)

<!-- Reviewable:end -->

pcwalton committed:
layout: Store viewport and screen size separately.

Fixes a bug whereby all nodes would get unconditionally reflowed on
every layout event if the page set a viewport.

pcwalton committed:
layout: Dump damage as well as part of the flow tree debug spew.

pcwalton committed:
layout: Minor whitespace cleanup.

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify layout code, but no reftests are modified. Please consider adding a reftest!

mbrubeck commented:
Reviewed 1 of 1 files at r1, 2 of 2 files at r2, 4 of 4 files at r3, 1 of 1 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

<sup>**[components/layout/display_list_builder.rs, line 1764 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8299#-K29N11Y0HjfL-l920Qe)** ([raw file](https://github.com/servo/servo/blob/596ecb7651e43a6fcb6caf5592fe83c3f4e1a514/components/layout/display_list_builder.rs#L1764)):</sup>
Would it make sense to put this in traversal::BuildDisplayList::process` instead (because that module deals more with restyle damage, and because then we don't need to ensure that each separate `build_display_list` implementation executes this)?
<!-- Reviewable comment -K29N11Y0HjfL-l920Qf:-1044873723 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8299)
<!-- Sent from Reviewable.io -->

pcwalton commented:
Updated. Added a new commit that I mistakenly stripped out: "layout: Don't repaint if the root flow doesn't have the `REPAINT` bit set."

r? @mbrubeck

frewsxcv commented:
Is adding a regression test for this non-trivial?

bholley commented:
Regression tests for these kinds of layout optimizations are pretty hard, because the optimizations are supposed to be transparent to content (i.e. you only skip things would be unobservable). We'd need hooks into the engine to do that kind of testing - might be worth doing, but almost certainly would require doing stuff outside the bounds of wpt.

mbrubeck commented:
Reviewed 3 of 3 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8299)
<!-- Sent from Reviewable.io -->

mbrubeck commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit dea9877 has been approved by `mbrubeck`

<!-- @bors-servo r=mbrubeck dea9877a86166eb536689e42052e910f58e7be13 -->

bors-servo commented:
:hourglass: Testing commit dea9877 with merge 7b1621d...

bors-servo commented:
:broken_heart: Test failed - [mac-dev-ref-unit](http://build.servo.org/builders/mac-dev-ref-unit/builds/186)

jdm commented:
```
failures:
    viewport_meta.html == viewport_rule_ref.html
    viewport_rule.html == viewport_rule_ref.html
```

pcwalton commented:
@bors-servo: r=mbrubeck

Fixed test failure.

bors-servo commented:
:pushpin: Commit 0d56154 has been approved by `mbrubeck`

<!-- @bors-servo r=mbrubeck 0d561540da2b133bb30d858047ce179e788942fb -->

bors-servo commented:
:hourglass: Testing commit 0d56154 with merge b138a20...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/184)

pcwalton commented:
Ugh. This is going to be one of *those* layout PRs.

jdm commented:
```
failures:
    iframe/hide_and_show.html == iframe/hide_and_show_ref.html
    iframe/multiple_external.html == iframe/multiple_external_ref.html
    iframe/positioning_margin.html == iframe/positioning_margin_ref.html
    iframe/simple.html == iframe/simple_ref.html
    iframe/simple_inline_default.html == iframe/simple_inline_default_ref.html
    iframe/simple_inline_height.html == iframe/simple_inline_height_ref.html
    iframe/simple_inline_min.html == iframe/simple_inline_min_ref.html
    iframe/simple_inline_width.html == iframe/simple_inline_width_ref.html
    iframe/simple_inline_width_percentage.html == iframe/simple_inline_width_percentage_ref.html
    iframe/stacking_context.html == iframe/stacking_context_ref.html
    iframe/stacking_context_position_a.html == iframe/stacking_context_position_ref.html
```

pcwalton commented:
I can't reproduce these failures on Mac. Will have to try on Linux.

pcwalton commented:
@bors-servo: r=mbrubeck

Pulled out the commit that I think might be causing the failures pending further investigation. Going to try to land the rest.

bors-servo commented:
:pushpin: Commit 2a5fe6e has been approved by `mbrubeck`

<!-- @bors-servo r=mbrubeck 2a5fe6e6b827a1281140f1bfae0316f745aa3472 -->

bors-servo commented:
:hourglass: Testing commit 2a5fe6e with merge f81a597...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/188)

eefriedman commented:
```
failures:
    iframe/bg_color.html == iframe/bg_color_ref.html
    iframe/hide_and_show.html == iframe/hide_and_show_ref.html
    iframe/multiple_external.html == iframe/multiple_external_ref.html
    iframe/positioning_margin.html == iframe/positioning_margin_ref.html
    iframe/simple.html == iframe/simple_ref.html
    iframe/simple_inline_height.html == iframe/simple_inline_height_ref.html
    iframe/simple_inline_max.html == iframe/simple_inline_max_ref.html
    iframe/simple_inline_min.html == iframe/simple_inline_min_ref.html
    iframe/simple_inline_width.html == iframe/simple_inline_width_ref.html
    iframe/simple_inline_width_height.html == iframe/simple_inline_width_height_ref.html
    iframe/size_attributes.html == iframe/size_attributes_ref.html
    iframe/stacking_context.html == iframe/stacking_context_ref.html
    iframe/stacking_context_position_a.html == iframe/stacking_context_position_ref.html
```

pcwalton commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 6633773 with merge a4b67bb...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/190), [gonk](http://build.servo.org/builders/gonk/builds/198), [linux-dev](http://build.servo.org/builders/linux-dev/builds/191), [linux-rel](http://build.servo.org/builders/linux-rel/builds/209), [mac-dev-ref-unit](http://build.servo.org/builders/mac-dev-ref-unit/builds/204), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/201), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/214)

pcwalton commented:
@bors-servo: r=mbrubeck

Passed try

bors-servo commented:
:pushpin: Commit 6633773 has been approved by `mbrubeck`

<!-- @bors-servo r=mbrubeck 6633773a589d87e81d2cf539983004d6ba957c68 -->

bors-servo commented:
:hourglass: Testing commit 6633773 with merge 3fb2a35...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/193), [gonk](http://build.servo.org/builders/gonk/builds/200), [linux-dev](http://build.servo.org/builders/linux-dev/builds/194), [linux-rel](http://build.servo.org/builders/linux-rel/builds/212), [mac-dev-ref-unit](http://build.servo.org/builders/mac-dev-ref-unit/builds/207), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/203), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/216)

bors-servo commented:
:eyes: Test was successful, but fast-forwarding failed: 422 Reference already exists

pcwalton commented:
@bors-servo: retry

bors-servo commented:
:hourglass: Testing commit 6633773 with merge 36c5dd4...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/199), [gonk](http://build.servo.org/builders/gonk/builds/206), [linux-dev](http://build.servo.org/builders/linux-dev/builds/201), [linux-rel](http://build.servo.org/builders/linux-rel/builds/219), [mac-dev-ref-unit](http://build.servo.org/builders/mac-dev-ref-unit/builds/214), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/209), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/222)
